File: 01/main.go
package main

import (
	"aoc-2021/util"
	"fmt"
	"path"
)

func partOne(nums []int) int {
	increases := 0
	prev := nums[0]

	for _, v := range nums[1:] {
		if v > prev {
			increases++
		}
		prev = v
	}

	return increases
}

func partTwo(nums []int) int {
	increases := 0

	sum := func(nums []int) int {
		res := 0
		for _, v := range nums {
			res += v
		}

		return res
	}

	for i := range nums[:len(nums)-3] {
		curr := nums[i : i+3]
		next := nums[i+1 : i+4]

		if sum(next) > sum(curr) {
			increases++
		}
	}

	return increases
}

func main() {
	nums := util.ReadNumbers(path.Join("01", "input.txt"))

	fmt.Println(partOne(nums))
	fmt.Println(partTwo(nums))
}

File: 01/main.js
const fs = require('fs');
const path = require('path');

const readInput = (file = '') => {
  const fileContent = fs.readFileSync(file, 'utf8');

  return fileContent;
};

const partOne = (input = '') => {
  // part one
  return input
    .split('\n')
    .map((i) => Number(i))
    .reduce((acc, curr, i, arr) => {
      return arr[i + 1] > arr[i] ? acc + 1 : acc;
    }, 0);
};

const partTwo = (input = '') => {
  // part two
  return input
    .split('\n')
    .map((i) => Number(i))
    .reduce((acc, curr, i, arr) => {
      const sumA = arr.slice(i, i + 3).reduce((acc, a) => acc + a, 0);
      const sumB = arr.slice(i + 1, i + 4).reduce((acc, a) => acc + a, 0);

      return sumB > sumA ? acc + 1 : acc;
    }, 0);
};

const input = readInput(path.join(__dirname, 'input.txt'));
console.log(partOne(input));
console.log(partTwo(input));
